src/model/openapi31.ts

```diff
@@ -292,7 +292,7 @@ export interface SchemaObject extends ISpecificationExtension {
     anyOf?: (SchemaObject | ReferenceObject)[];
     not?: SchemaObject | ReferenceObject;
     items?: SchemaObject | ReferenceObject;
-    properties?: { [propertyName: string]: SchemaObject | ReferenceObject };
+    properties?: Record<string, SchemaObject | ReferenceObject>;
```

Comment:
I would suggest : `Record<string, SchemaObject | ReferenceObject | undefined>`

Reply:
you can use `Partial<Record>` instead of including `undefined`

Reply:
I don't think the value of any field in `properties` can ever be `undefined`.

In YAML spec files, If no value is assigned to a field in `properties`, Swagger automatically treats the field as of type `any`.

<img width="1507" alt="image" src="https://github.com/metadevpro/openapi3-ts/assets/47082523/5d80c25a-dd55-45a7-92ca-d64aa932fb13">


However, in JSON format, it's not allowed to assign an undefined value to a filed in `properties`

<img width="1493" alt="image" src="https://github.com/metadevpro/openapi3-ts/assets/47082523/f6b703e6-01f2-4ec0-921e-8f723ce00bbe">

Reply:
In the worst case, it'll be an empty object, which again is a valid `SchemaObject`

Reply:
@remidewitte do you think `undefined` would be needed here?

Reply:
> But that condition can be implemented differently, without involving undefined at all:
```js
yaml.stringify({
  ...(isIncluded() ? {property: "something"} : {})
});
```

And `{}` is a valid `SchemaObject` (although an invalid `ReferenceObject`), so `Record<string, SchemaObject | ReferenceObject>` should work fine as the type for `properties`.

Reply:
@remidewitte @RobinTail what do you reckon will be the optimal approach here?

Reply:
I recommend to keep focus of this PR on its originally declared problem:
making keys of type `string | number`.

`undefined` was not there and it's a feature that may or may not be required, so it could be another dedicated PR, if needed, @vaibhavrajsingh2001 , @remidewitte

Reply:
The thing is in a programmatic usage : if you try to get a property that is not defined at all, it should tell that it may not defined thanks to typescript...

```yaml
type: object
properties: 
  one: 
     type: string
```

```ts 
const obj  : SchemaObject = ...;
obj.properties['one']; // It can be undefined and it is not
obj.properties['two']; // It can  be undefined and it is
```

I am ok to address this in a separate PR if you feel like it.

Reply:
I believe it's another problem related to the strictness level of Typescript, @remidewitte 
There is a setting for that in tsconfig:

https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess

By default it's disabled, @remidewitte

Reply:
My understanding of [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess) is (but I might be wrong) : 
```
obj.properties['two']; // OK
obj.properties.two; // Error
```

Reply:
> Turning on noUncheckedIndexedAccess will add undefined to any un-declared field in the type

@remidewitte

Reply:
with `noUncheckedIndexedAccess` enabled, @remidewitte 

https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true#code/MYewdgzgLgBCBGArAXDASgU1AJwCYB5psBLMAcwBoYBvbDAM1SNLIF8A+GAXhtYG4AUANCRYUDNG5wkAOnHQZdenyA

without it:

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBCBGArAXDASgU1AJwCYB5psBLMAcwBoYBvbDAM1SNLIF8A+GAXhtYG4AUANCRYUDNG5wkAOnHQZdenyA

Reply:
So, should I update the types to `Partial<Record<string, SchemaObject | ReferenceObject>>`?

Reply:
@vaibhavrajsingh2001 , Taking into account the concept and the goal of this package:

> TypeScript library to help building OpenAPI 3.x compliant API contracts.

**I'd say: there are more benefits in `Partial<>` than any harm.**

Under the hood `Partial<>` makes all properties optional `?:`, while does not explicitly welcome `undefined`,
which can be assigned, but still there is no harm of that — it will not go further (JSON and YAML stringifiers remove it). It's helpful both for conditional writing and for avoiding runtime errors due while reading, like in the case of @remidewitte

Reply:
Updated the PR @RobinTail @remidewitte